### PR TITLE
TiKV failover

### LINKED ISSUE #58: TiKV automatic failover
When a TiKV peer's state is not `Up` for a while, operator should just increase the replicas to add a new TiKV peer, that is all.

port of #47

### PR DESCRIPTION
**Pull request by weekface**
This PR fixes #58, adds TiKV failover feature to TiDB Operator.

@tennix @onlymellb @xiaojingchen @gregwebs PTAL

### CODE REVIEW

File: pkg/manager/member/tikv_member_manager.go [hunk 1]
```diff
@@ -185,6 +191,18 @@ func (tkmm *tikvMemberManager) syncStatefulSetForTidbCluster(tc *v1alpha1.TidbCl
 		}
 	}
 
+	if tkmm.autoFailover {
+		if allTiKVStoressAreReady(tc) {
+			if tc.Status.TiKV.FailureStores != nil {
+				tkmm.tikvFailover.Recover(tc)
+			}
+		} else if int(tc.Spec.TiKV.Replicas) == int(tc.Status.TiKV.StatefulSet.Replicas) {
```
Comment (tennix) [inline pkg/manager/member/tikv_member_manager.go]: Unnecessary cast `replicas` to `int`

File: pkg/manager/member/tikv_failover_test.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"time"
+
+	. "github.com/onsi/gomega"
```
Comment (tennix) [inline pkg/manager/member/tikv_failover_test.go]: Add revive instruction for `dot-import`.
You can choose whatever you like to follow revive linter suggests.
`retool do revive -formatter friendly -config revive.toml pkg/...`

File: pkg/manager/member/tikv_failover_test.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+package member
+
+import (
+	"fmt"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"time"
```
Comment (tennix) [inline pkg/manager/member/tikv_failover_test.go]: Could you separate the standard library and external library?

File: pkg/manager/member/tikv_failover_test.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		tc := newTidbClusterForPD()
+		test.update(tc)
+		tikvFailover, fakePDControl := newFakeTiKVFailover()
+		pdClient := controller.NewFakePDClient()
+		fakePDControl.SetPDClient(tc, pdClient)
+
+		pdClient.AddReaction(controller.GetConfigActionType, func(action *controller.Action) (interface{}, error) {
+			if test.getCfgErr {
+				return nil, fmt.Errorf("get config failed")
+			} else {
```
Comment (tennix) [inline pkg/manager/member/tikv_failover_test.go]: No need to add `else` block.

File: pkg/apis/pingcap.com/v1alpha1/types.go [hunk 5]
```diff
@@ -194,15 +193,11 @@ type TiDBMember struct {
...
-	Stores      TiKVStores              `json:"stores,omitempty"`
-}
-
-// TiKVStores is all TiKV Stores, contain stores which state are tombstone
-type TiKVStores struct {
-	CurrentStores   map[string]TiKVStore `json:"currentStores,omitempty"`
-	TombStoneStores map[string]TiKVStore `json:"tombStoneStores,omitempty"`
+	Phase           MemberPhase                 `json:"phase,omitempty"`
+	StatefulSet     *apps.StatefulSetStatus     `json:"statefulSet,omitempty"`
+	Stores          map[string]TiKVStore        `json:"stores,omitempty"`
```
[the remark this answers is not in the document]
Comment (weekface) [reply pkg/apis/pingcap.com/v1alpha1/types.go]: `Stores` contain failure stores, but don't contain tombstone stores.

**Pull request merged**